File: src/indexer.ts
// Files are truth: the scan is the input, the DB is the output. Dirtiness is
// decided by content hash, never by the DB's own bookkeeping.
import type { Database } from "bun:sqlite";
import { lstatSync, readdirSync, type Stats } from "node:fs";
import { join, relative } from "node:path";
import { linkTarget, parseNote, qualifyLinks, type Note } from "./note";
import { confinedPath, writeAtomic } from "./gate";
import { resetVectors, vectorsStale } from "./db";
import { l2normalize, type Embedder } from "./embed";

/**
 * One stem collision's auto-qualify outcome (DESIGN.md § Data model): the bare
 * `[[stem]]` links that still unambiguously meant the incumbent, rewritten to
 * its path-qualified form. `target` is null when the incumbent sits at the
 * vault root — it has no qualified form, so nothing is rewritten and every
 * linker is reported. `skipped` also carries linkers edited mid-flight (hash
 * mismatch — never clobbered) and the remainder past the cap; all of them fall
 * back to doctor's ambiguous report. The invariant is *never guesses*, not
 * *never ambiguous*.
 */
export type Qualified = {
  stem: string;
  /** the incumbent's path-qualified link target, or null for a root incumbent */
  target: string | null;
  /** linking notes whose bare links were rewritten */
  rewritten: string[];
  /** linking notes left alone: root incumbent, edited mid-flight, or over the cap */
  skipped: string[];
};

export type IndexStats = {
  added: number;
  updated: number;
  removed: number;
  unchanged: number;
  /** the embedding model or dims changed, so every note was re-embedded */
  reembedded: boolean;
  /** stem collisions this pass created, and the bare links qualified for them */
  qualified: Qualified[];
};

/**
 * Vault-relative paths of every `.md` file, sorted. Dot-directories (`.vault`,
 * `.git`, `.obsidian`, …) are skipped and symlinks are never followed — a link
 * out of the vault is not a note, and a link back in is not a second one.
 */
export function scanVault(root: string): string[] {
  const out: string[] = [];
  const walk = (dir: string): void => {
    for (const entry of readdirSync(dir, { withFileTypes: true })) {
      if (entry.name.startsWith(".") || entry.isSymbolicLink()) continue;
      const abs = join(dir, entry.name);
      if (entry.isDirectory()) walk(abs);
      // Separators are normalized to `/`: a stored path is compared against
      // `/`-written wikilink targets, and `relative` yields `\` on Windows.
      else if (entry.isFile() && entry.name.endsWith(".md")) {
        out.push(relative(root, abs).replaceAll("\\", "/"));
      }
    }
  };
  walk(root);
  return out.sort();
}

/**
 * Is this vault-relative path one `scanVault` would index — a `.md` file with
 * no dot-directory (or dot-file) in it? The same rule, applied to a path we
 * were handed rather than one we walked to: `fs.watch` reports paths, not
 * directory entries, so the watcher cannot check for symlinks the way the scan
 * does (one that slips in is purged by the next `doctor`).
 */
export function isNotePath(rel: string): boolean {
  return rel.endsWith(".md") && !rel.split(/[/\\]/).some((segment) => segment.startsWith("."));
}

/**
 * The `lstat` of a path that really holds a note, or null. **Only a regular
 * file is a note** (DESIGN.md § Layout): a path that is gone, has become a
 * directory, or has become a symlink — which the scan skips, and which can
 * point clean out of the vault — holds no note. Decided *before* the read and
 * shared, so the indexer's deletion rule and `get`'s null cannot drift apart.
 */
export function noteEntry(root: string, rel: string): Stats | null {
  const entry = lstatSync(join(root, rel), { throwIfNoEntry: false });
  return entry?.isFile() ? entry : null;
}

/**
 * Read and parse one note, or null if it is no longer there. A human can
 * delete a file between the scan and the read; that is an ordinary event in a
 * files-are-truth vault, not a reason to abort the run. Any other read error
 * (permissions, I/O) still throws.
 */
export async function readNote(root: string, rel: string): Promise<Note | null> {
  const raw = await readRaw(root, rel);
  return raw === null ? null : parseNote(raw, rel);
}

/** The file's text, or null if it is no longer there (see `readNote`). */
export async function readRaw(root: string, rel: string): Promise<string | null> {
  try {
    return await Bun.file(join(root, rel)).text();
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === "ENOENT") return null;
    throw e;
  }
}

/** Hash-diff the whole vault against the index and write only what changed. */
export async function reindex(
  db: Database,
  root: string,
  embedder: Embedder,
): Promise<IndexStats> {
  // Every path the files know about plus every path the index knows about: the
  // ones only the index has are deletions, and `indexPaths` purges them.
  const indexed = (db.query(`select path from notes`).all() as { path: string }[]).map(
    (r) => r.path,
  );
  const stats = await indexPaths(db, root, embedder, [...scanVault(root), ...indexed]);
  // Unconditionally, unlike the pass inside `indexPaths`: an index written by
  // an older version holds `to_id` values *that* rule produced, and no file has
  // to change for them to be wrong. The whole-vault entry point is where a
  // changed rule gets applied — otherwise a link that only the new rule can
  // resolve stays broken until someone edits a note or runs `--rebuild`. The
  // watcher's no-op pass stays writeless because it does not come through here.
  resolveEdges(db);
  return stats;
}

/**
 * Hash-diff exactly these vault-relative paths and write only what changed; a
 * path whose file is gone is purged. `reindex` passes the whole vault, the
 * watcher passes the handful of paths that just changed — one code path, so a
 * watched vault and a rebuilt one cannot end up with different rows.
 * ponytail: reads the whole `notes` table even for one path (three columns of a
 * hand-written vault). Restrict it to the paths asked for if that ever shows up.
 *
 * When this pass *creates* a stem collision — a newly indexed note's stem
 * matches exactly one note the index already knew about — every bare
 * `[[stem]]` link in the vault still unambiguously means that incumbent, and
 * only this moment can know it: once both notes are indexed, nothing records
 * which was first. Those links are rewritten to the incumbent's path-qualified
 * form after the write transaction commits (no file I/O inside it — a rollback
 * cannot unwrite a file), check-and-write gated, capped at `qualifyCap` per
 * collision, and reported in `stats.qualified`. The rewritten paths re-enter
 * this function once; their stems do not change, so no further collisions and
 * no recursion.
 */
export async function indexPaths(
  db: Database,
  root: string,
  embedder: Embedder,
  rels: Iterable<string>,
  qualifyCap = QUALIFY_CAP,
): Promise<IndexStats> {
  // Only asked here, never acted on: the vectors this invalidates are the ones
  // the vault searches with until the replacements exist, and `embed` below is
  // a network call that fails for ordinary reasons. The drop happens in the
  // write transaction, with the new vectors in hand.
  const reembedded = vectorsStale(db, embedder);
  const rows = db.query(`select id, path, hash, slug from notes`).all() as {
    id: number;
    path: string;
    hash: string;
    slug: string;
  }[];
  const indexed = new Map(rows.map((r) => [r.path, r]));
  const embedded = new Set(
    (db.query(`select note_id from vector_meta`).all() as { note_id: number }[]).map(
      (r) => r.note_id,
    ),
  );

  const dirty: { note: Note; mtime: number; id: number | undefined }[] = [];
  // Rows whose path is no longer a note on disk — see the entry check below.
  const gone: number[] = [];
  let unchanged = 0;
  for (const rel of new Set(rels)) {
    const row = indexed.get(rel);
    // What is at the path decides, and it is decided *before* the read: only a
    // regular file is a note. A path that is gone, has become a directory, or
    // has become a symlink — which the scan skips, and which can point clean
    // out of the vault — is a deletion. Reading first would index a symlink's
    // target from outside the vault into a row nothing can ever purge (the
    // scan never lists it again, so doctor calls it missing forever), and would
    // throw EISDIR out of both reindex and doctor on a path turned directory,
    // leaving no way to repair the vault at all.
    const entry = noteEntry(root, rel);
    // Still null-checked after the read: a human may delete the file between
    // the two syscalls, which is an ordinary event, not a failure.
    const note = entry === null ? null : await readNote(root, rel);
    if (note === null) {
      if (row) gone.push(row.id);
      continue;
    }
    // A row with no vector is half-indexed (interrupted run) — redo it.
    if (row && row.hash === note.hash && !reembedded && embedded.has(row.id)) {
      unchanged++;
      continue;
    }
    dirty.push({ note, mtime: Math.floor(entry?.mtimeMs ?? 0), id: row?.id });
  }

  // Detect, before the write transaction, from the snapshot above: a new note
  // (`id === undefined`) whose stem exactly one pre-existing note still holds —
  // "still": a row in `gone` is a rename or deletion, not an incumbent. Zero
  // holders is no collision; two or more means the stem was already ambiguous
  // and there is nothing left to protect. Deduped per stem: two newcomers
  // colliding with the same incumbent are one rewrite.
  const goneSet = new Set(gone);
  const collisions = new Map<string, string>(); // stem → incumbent path
  for (const d of dirty) {
    if (d.id !== undefined) continue;
    const holders = rows.filter((r) => r.slug === d.note.slug && !goneSet.has(r.id));
    if (holders.length === 1) collisions.set(d.note.slug, holders[0]!.path);
  }

  const vectors = dirty.length
    ? await embedder.embed(dirty.map((d) => `${d.note.title}\n\n${d.note.body}`))
    : [];
  // Catch a misbehaving provider here, not as a raw sqlite-vec failure halfway
  // through the write transaction.
  if (vectors.length !== dirty.length) {
    throw new Error(
      `embedder ${embedder.model} returned ${vectors.length} vectors for ${dirty.length} texts`,
    );
  }
  const wrong = vectors.find((v) => v.length !== embedder.dims);
  if (wrong) {
    throw new Error(
      `embedder ${embedder.model} returned a vector of width ${wrong.length}, expected ${embedder.dims}`,
    );
  }

  const upsert = db.query(`insert into notes
      (path, slug, title, type, hash, frontmatter, superseded_by, mtime, malformed)
      values (?, ?, ?, ?, ?, ?, ?, ?, ?)
    on conflict(path) do update set
      slug = excluded.slug, title = excluded.title, type = excluded.type,
      hash = excluded.hash, frontmatter = excluded.frontmatter,
      superseded_by = excluded.superseded_by, mtime = excluded.mtime,
      malformed = excluded.malformed
    returning id`);

  db.transaction(() => {
    // Now: the replacements are embedded and validated, and a failure from here
    // on rolls the drop back with everything else.
    resetVectors(db, embedder, reembedded);
    for (const [i, { note, mtime }] of dirty.entries()) {
      const supersededBy = note.frontmatter["superseded_by"];
      const { id } = upsert.get(
        note.path,
        note.slug,
        note.title,
        note.type ?? null,
        note.hash,
        JSON.stringify(note.frontmatter),
        typeof supersededBy === "string" ? supersededBy : null,
        mtime,
        note.malformedFrontmatter ? 1 : 0,
      ) as { id: number };

      db.run(`delete from notes_fts where rowid = ?`, [id]);
      db.run(`insert into notes_fts (rowid, title, body) values (?, ?, ?)`, [
        id,
        note.title,
        note.body,
      ]);
      db.run(`delete from edges where from_id = ?`, [id]);
      // `to_slug` holds the link as written — a bare stem or a path; see
      // `resolveEdges` for how each is resolved.
      for (const target of note.links) {
        db.run(`insert or ignore into edges (from_id, to_slug) values (?, ?)`, [id, target]);
      }
      db.run(`delete from vectors where note_id = ?`, [id]);
      // A note with no tokens the embedder recognises (CJK, emoji, empty)
      // yields an all-zero vector, which has no direction — cosine distance
      // against it is NaN and would poison KNN. Skip the row; the note stays
      // findable through FTS. The meta row still records the attempt, so the
      // note does not look half-indexed on every later pass.
      const emb = l2normalize(vectors[i]!);
      if (emb.some((x) => x !== 0)) {
        db.run(`insert into vectors (note_id, emb) values (?, ?)`, [id, emb]);
      }
      db.run(`insert or replace into vector_meta (note_id, model, dims) values (?, ?, ?)`, [
        id,
        embedder.model,
        embedder.dims,
      ]);
    }
    for (const id of gone) purgeNote(db, id);
    // Resolution depends only on the note set, so an unchanged pass is a no-op
    // — and must stay one, or "nothing changed" would still dirty the file.
    if (dirty.length || gone.length) resolveEdges(db);
  })();

  // Rewrite after commit: the moment both notes are indexed is the only one
  // that knows which was the incumbent, and the only file writes `indexPaths`
  // makes. Between the commit above and the re-entry below, the qualified
  // edges briefly read `to_id = null` — the documented window.
  const qualified: Qualified[] = [];
  const rewritten: string[] = [];
  // A bare link *inside* a colliding newcomer has no settled meaning — it was
  // written when its own note was the collision — so it is left to doctor.
  const newPaths = new Set(dirty.filter((d) => d.id === undefined).map((d) => d.note.path));
  for (const [stem, incumbent] of collisions) {
    // A root incumbent's path minus `.md` *is* the bare stem: no qualified
    // form, so its linkers are reported, never rewritten (DESIGN.md § Layout).
    const target = incumbent.includes("/") ? linkTarget(incumbent) : null;
    const linkers = db
      .query(
        `select n.path, n.hash from edges e join notes n on n.id = e.from_id
         where e.to_slug = ? order by n.path`,
      )
      .all(stem) as { path: string; hash: string }[];
    const entry: Qualified = { stem, target, rewritten: [], skipped: [] };
    for (const { path, hash } of linkers) {
      if (newPaths.has(path)) continue;
      if (target === null || entry.rewritten.length >= qualifyCap) {
        entry.skipped.push(path);
        continue;
      }
      const abs = confinedPath(root, path); // the index is derived data, not a trusted path source
      const raw = await readRaw(root, path);
      if (raw === null) continue; // indexed but gone: files are truth, the row is stale
      // Check-and-write: the file must still be what the index read it at — a
      // human's mid-flight edit is skipped, never clobbered, like supersede's
      // `unmarked`.
      if (parseNote(raw, path).hash !== hash) {
        entry.skipped.push(path);
        continue;
      }
      // In-place over the raw text (body region only): every byte this rewrite
      // does not mean — frontmatter, CRLF, a BOM — survives verbatim.
      const out = qualifyLinks(raw, stem, target);
      if (out === raw) continue; // nothing the parser calls a link actually matched
      await writeAtomic(abs, out);
      entry.rewritten.push(path);
      rewritten.push(path);
    }
    qualified.push(entry);
  }
  // Re-enter on just the rewritten paths so the index never lags a write we
  // made ourselves. Bounded structurally: none of them is new to the index, so
  // the pass detects no collision and recurses no further.
  if (rewritten.length > 0) await indexPaths(db, root, embedder, rewritten);

  return {
    added: dirty.filter((d) => d.id === undefined).length,
    updated: dirty.filter((d) => d.id !== undefined).length,
    removed: gone.length,
    unchanged,
    reembedded,
    qualified,
  };
}

/**
 * Rewrites per collision before the remainder is reported instead (#29): the
 * rewrite is deterministic, so this is a generous fuse against a pathological
 * vault, not a tuning knob. Injectable via `indexPaths` for the tests.
 */
export const QUALIFY_CAP = 500;

/** Drop every derived row for a note whose file is gone. */
export function purgeNote(db: Database, id: number): void {
  db.run(`delete from notes where id = ?`, [id]);
  db.run(`delete from notes_fts where rowid = ?`, [id]);
  db.run(`delete from edges where from_id = ?`, [id]);
  db.run(`delete from vectors where note_id = ?`, [id]);
  db.run(`delete from vector_meta where note_id = ?`, [id]);
}

/**
 * Resolve every wikilink, namespace-aware (DESIGN.md § Data model). A target
 * carrying a `/` is a **path**: it matches one note's vault-relative path minus
 * its `.md`, exactly, so `[[customers/acme]]` and `[[vendors/acme]]` are two
 * different notes. A target without one is a **bare stem**: it resolves only
 * when exactly one note in the vault carries that filename stem — several
 * candidates leave `to_id` null (ambiguous) rather than picking a winner, which
 * is the only answer that cannot silently point at the wrong note.
 *
 * Both branches are SQL against paths the scan already found: a link is never
 * joined onto the filesystem, so `[[../../etc/passwd]]` is not a traversal, it
 * is a string that matches no row.
 *
 * Recomputed wholesale because adding or removing *any* note can flip a link
 * either way, including links in notes that did not themselves change.
 * ponytail: rewrites every edge on any change — scope it to the touched targets
 * if a vault ever gets big enough for that to show up.
 */
export function resolveEdges(db: Database): void {
  db.run(`update edges set to_id = case
    when instr(to_slug, '/') > 0
      then (select n.id from notes n where n.path = edges.to_slug || '.md')
      else (select min(n.id) from notes n where n.slug = edges.to_slug having count(*) = 1)
  end`);
}

File: src/wilcus_vault/indexer.py
"""Files are truth: the scan is the input, the database is the output.

Dirtiness is decided by content hash, never by the database's own bookkeeping,
and one function (`index_paths`) writes every index row.
"""

import os
import sqlite3
import stat
from collections.abc import Iterable
from dataclasses import dataclass, field
from pathlib import Path

from .db import reset_vectors, transaction, vectors_stale
from .embed import Embedder
from .index_rows import Dirty, purge_note, resolve_edges, write_note
from .note import Note, parse_note
from .qualify import Qualified, detect_collisions, qualify_collisions
from .term import VaultError, printable

# Rewrites per collision before the rest is left to doctor. A generous fuse
# against a pathological vault, not a tuning knob.
QUALIFY_CAP = 500


@dataclass
class IndexStats:
    added: int = 0
    updated: int = 0
    removed: int = 0
    unchanged: int = 0
    reembedded: bool = False  # the model or dims changed, so every note was re-embedded
    qualified: list[Qualified] = field(default_factory=list)  # stem collisions this pass handled
    # The re-index of rewritten linkers failed; their rows lag the files until
    # the next pass. Reported, not raised, so the stats still say what changed.
    index_error: str | None = None


def scan_vault(root: str | Path) -> list[str]:
    """Vault-relative paths of every `.md` file, sorted. Dot-directories are
    skipped and symlinks are never followed."""
    root = Path(root)
    out = []
    for dirpath, dirnames, filenames in os.walk(root):
        here = Path(dirpath)
        dirnames[:] = [d for d in dirnames if not d.startswith(".") and not (here / d).is_symlink()]
        for name in filenames:
            path = here / name
            if name.startswith(".") or not name.endswith(".md") or path.is_symlink():
                continue
            if path.is_file():
                out.append(path.relative_to(root).as_posix())
    return sorted(out)


def is_note_path(rel: str) -> bool:
    """Would the scan index this path: a `.md` file with no dot-segment in it?"""
    segments = rel.replace("\\", "/").split("/")
    return rel.endswith(".md") and not any(s.startswith(".") for s in segments)


def note_entry(root: str | Path, rel: str) -> os.stat_result | None:
    """The lstat of a path that really holds a note, or None. Only a regular file
    is a note: a path that is gone, a directory, or a symlink holds none.

    Any OSError is that same answer — a segment that is a file not a directory,
    a name too long for the filesystem, a directory we may not read — so a
    caller asking "is there a note here" never has to catch errno itself.
    """
    try:
        entry = os.lstat(Path(root) / rel)
    except OSError:
        return None
    return entry if stat.S_ISREG(entry.st_mode) else None


def read_raw(root: str | Path, rel: str) -> str | None:
    """The file's text, or None if it is no longer there. Other errors raise."""
    try:
        with (Path(root) / rel).open(encoding="utf-8", errors="replace", newline="") as f:
            return f.read()
    except FileNotFoundError:
        return None


def read_note(root: str | Path, rel: str) -> Note | None:
    raw = read_raw(root, rel)
    return None if raw is None else parse_note(raw, rel)


async def reindex(db: sqlite3.Connection, root: str | Path, embedder: Embedder) -> IndexStats:
    """Hash-diff the whole vault against the index and write only what changed."""
    # Every path the files know plus every path the index knows: the ones only
    # the index has are deletions, and index_paths purges them.
    indexed = [r["path"] for r in db.execute("select path from notes")]
    stats = await index_paths(db, root, embedder, [*scan_vault(root), *indexed])
    # Unconditional, unlike inside index_paths: an index written by an older
    # version may hold to_id values an older resolution rule produced.
    resolve_edges(db)
    return stats


async def index_paths(
    db: sqlite3.Connection,
    root: str | Path,
    embedder: Embedder,
    rels: Iterable[str],
    qualify_cap: int = QUALIFY_CAP,
) -> IndexStats:
    """Hash-diff exactly these paths and write only what changed; a path whose
    file is gone is purged. When the pass creates a stem collision, bare links
    to the incumbent are rewritten to its qualified form after the commit."""
    root = Path(root)
    # Asked here, acted on inside the write transaction with the new vectors in hand.
    reembedded = vectors_stale(db, embedder)
    rows = db.execute("select id, path, hash, slug from notes").fetchall()
    by_path = {r["path"]: r for r in rows}
    embedded = {r["note_id"] for r in db.execute("select note_id from vector_meta")}

    dirty: list[Dirty] = []
    gone: list[int] = []
    unchanged = 0
    for rel in dict.fromkeys(rels):
        row = by_path.get(rel)
        # What is at the path is decided before the read: only a regular file
        # is a note. A directory or symlink there is a deletion.
        entry = note_entry(root, rel)
        note = None if entry is None else read_note(root, rel)
        if note is None or entry is None:
            if row is not None:
                gone.append(row["id"])
            continue
        # A row with no vector is half-indexed (interrupted run): redo it.
        if row and row["hash"] == note.hash and not reembedded and row["id"] in embedded:
            unchanged += 1
            continue
        dirty.append(Dirty(note, int(entry.st_mtime * 1000), row["id"] if row else None))

    new_notes = [d.note for d in dirty if d.id is None]
    collisions = detect_collisions(root, rows, new_notes, set(gone))

    texts = [f"{d.note.title}\n\n{d.note.body}" for d in dirty]
    vectors = await embedder.embed(texts) if texts else []
    if len(vectors) != len(dirty):
        raise VaultError(
            f"embedder {embedder.model} returned {len(vectors)} vectors for {len(dirty)} texts"
        )
    for v in vectors:
        if len(v) != embedder.dims:
            raise VaultError(
                f"embedder {embedder.model} returned a vector of width {len(v)}, "
                f"expected {embedder.dims}"
            )

    with transaction(db):
        reset_vectors(db, embedder, reembedded)
        for d, vector in zip(dirty, vectors, strict=True):
            write_note(db, d, vector, embedder)
        for note_id in gone:
            purge_note(db, note_id)
        # Resolution depends only on the note set, so an unchanged pass stays writeless.
        if dirty or gone:
            resolve_edges(db)

    new_paths = {n.path for n in new_notes}
    qualified, rewritten, index_error = qualify_collisions(
        db, root, collisions, new_paths, qualify_cap
    )
    # Re-enter on the rewritten paths so the index never lags our own writes.
    # None of them is new, so no further collision is detected.
    if rewritten:
        try:
            await index_paths(db, root, embedder, rewritten)
        except Exception as e:
            index_error = index_error or printable(e)

    updated = {d.note.path for d in dirty if d.id is not None} | set(rewritten)
    return IndexStats(
        added=len(new_notes),
        updated=len(updated),
        removed=len(gone),
        unchanged=unchanged,
        reembedded=reembedded,
        qualified=qualified,
        index_error=index_error,
    )
